### PR TITLE
Update picard/collect*metrics

### PR DESCRIPTION
**Pull request by matthdsm**
For all picard/Collect*Metrics modules
- Add optional bam index
- Add meta to fasta, fai

And fixed the subworkflow

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by maxulysse** [comment]
@nf-core-bot fix linting pretty please :pray:

File: modules/nf-core/picard/collecthsmetrics/main.nf [hunk 1]
```diff
@@ -8,9 +8,9 @@ process PICARD_COLLECTHSMETRICS {
         'quay.io/biocontainers/picard:2.27.4--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(bam)
-    path fasta
-    path fai
+    tuple val(meta), path(bam), path(bai)
+    tuple val(meta2), path(fasta)
+    tuple val(meta2), path(fai)
```
Comment (maxulysse) [inline modules/nf-core/picard/collecthsmetrics/main.nf:13]:
```suggestion
    tuple val(meta3), path(fai)
```

File: modules/nf-core/picard/collecthsmetrics/meta.yml [hunk 2]
```diff
@@ -26,14 +26,28 @@ input:
         e.g. [ id:'test', single_end:false ]
   - bam:
       type: file
-      description: An aligned BAM/SAM file
+      description: An aligned BAM/CRAM/SAM file
       pattern: "*.{bam,sam}"
```
Comment (maxulysse) [inline modules/nf-core/picard/collecthsmetrics/meta.yml:30]:
```suggestion
      pattern: "*.{bam,cram,sam}"
```

File: modules/nf-core/picard/collecthsmetrics/meta.yml [hunk 3]
```diff
@@ -26,14 +26,28 @@ input:
...
+  - bai:
+      type: file
+      description: Optional aligned BAM/CRAM/SAM file index
+      pattern: "*.{bai,crai,sai}"
+  - meta2:
+      type: map
+      description: |
+        Groovy Map containing reference information
+        e.g. [ id:'genome' ]
```
Comment (maxulysse) [inline modules/nf-core/picard/collecthsmetrics/meta.yml:39]: we should add meta3 too

Comment (maxulysse) [reply modules/nf-core/picard/collecthsmetrics/meta.yml:39]: nevermind

File: modules/nf-core/picard/collecthsmetrics/meta.yml [hunk 4]
```diff
@@ -26,14 +26,28 @@ input:
...
   - fasta:
       type: file
       description: |
         A reference file to calculate dropout metrics measuring reduced representation of reads.
         Optional input.
-      pattern: "*.fasta"
+      pattern: "*.{fa,fasta,fna}"
+  - meta2:
```
Comment (maxulysse) [inline modules/nf-core/picard/collecthsmetrics/meta.yml:46]:
```suggestion
  - meta3:
```

File: modules/nf-core/picard/collecthsmetrics/main.nf (hunk 1 above)
Comment (matthdsm) [reply modules/nf-core/picard/collecthsmetrics/main.nf:13]: shouldn't the fasta and it's index have the same meta? It's like that in other subwfs I think

Comment (maxulysse) [reply modules/nf-core/picard/collecthsmetrics/main.nf:13]: they're different files.
I'd say the fasta can have a meta describing species + version + file, which then would be different from the fai, that could on top have info on the version of the tools that build it

Comment (maxulysse) [reply modules/nf-core/picard/collecthsmetrics/main.nf:13]: Plus, I'm concerned about collapsing meta maps, even if it should not matter much in that case

**Comment by maxulysse** [comment]
You should also update the subworkflow that depend on it: `subworkflows/nf-core/bam_qc_picard`

File: modules/nf-core/picard/collecthsmetrics/main.nf (hunk 1 above)
Comment (matthdsm) [reply modules/nf-core/picard/collecthsmetrics/main.nf]: Fixed it!

File: subworkflows/nf-core/bam_qc_picard/main.nf [hunk 5]
```diff
@@ -8,27 +8,27 @@ include { PICARD_COLLECTHSMETRICS       } from '../../../modules/nf-core/picard/
 
 workflow BAM_QC_PICARD {
     take:
-    ch_bam              // channel: [ val(meta), [ bam ]]
-    ch_fasta            // channel: [ fasta ]
-    ch_fasta_fai        // channel: [ fasta_fai ]
+    ch_bam_bai          // channel: [ val(meta), [ bam ], [bai]]
+    ch_fasta            // channel: [ val(meta2), fasta ]
+    ch_fasta_fai        // channel: [ val(meta2), fasta_fai ]
```
[the remark this answers is not in the document]
Comment (matthdsm) [reply subworkflows/nf-core/bam_qc_picard/main.nf:13]: Don't we need to correlate with the value in `meta.yml`?

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/bam_qc_picard/main.nf:13]: not here, here they're just meta maps, it's in the modules that they became meta2 or meta3

[the remark this answers is not in the document]
Comment (matthdsm) [reply subworkflows/nf-core/bam_qc_picard/main.nf:13]: fine by me!

**Comment by matthdsm** [comment]
tests passed! Ready for approval!

**Review by maxulysse — APPROVED** [review]

**Pull request merged**